compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -244,22 +244,20 @@ pub fn each_linked_rlib(
 
         fmts
     } else {
-        for combination in info.dependency_formats.iter().combinations(2) {
-            let (ty1, list1) = &combination[0];
-            let (ty2, list2) = &combination[1];
-            if list1 != list2 {
-                return Err(errors::LinkRlibError::IncompatibleDependencyFormats {
-                    ty1: format!("{ty1:?}"),
-                    ty2: format!("{ty2:?}"),
-                    list1: format!("{list1:?}"),
-                    list2: format!("{list2:?}"),
-                });
-            }
-        }
-        if info.dependency_formats.is_empty() {
-            return Err(errors::LinkRlibError::MissingFormat);
+        if let Some(((ty1, list1), (ty2, list2))) = info
+            .dependency_formats
+            .iter()
+            .tuple_combinations()
+            .find(|((_, list1), (_, list2))| list1 != list2)
+        {
+            return Err(errors::LinkRlibError::IncompatibleDependencyFormats {
+                ty1: format!("{ty1:?}"),
+                ty2: format!("{ty2:?}"),
+                list1: format!("{list1:?}"),
+                list2: format!("{list2:?}"),
+            });
         }
```

Comment:
This is a pre-existing issue: why does the code look at all possible combinations of 2 dependency formats? All that this loop does is ensure that all `list`s are the same. Simply iterating once should suffice.
```suggestion
        let dep_formats = info.dependency_formats.iter();
        let (ty1, list1) = dep_formats.next().ok_or(errors::LinkRlibError::MissingFormat)?;
        if let Some((ty2, list2)) = dep_formats.find(|(_, list2)| list1 != list2) {
            return Err(errors::LinkRlibError::IncompatibleDependencyFormats {
                ty1: format!("{ty1:?}"),
                ty2: format!("{ty2:?}"),
                list1: format!("{list1:?}"),
                list2: format!("{list2:?}"),
            });
        }
        list1
```
(This will probably change the output of some UI test slightly)

Reply:
Wow, nice catch, thanks!

Reply:
I'm not sure it'll change any UI test output, the earliest element with a different `list` should trigger the error, before and after this change. I think. I'll try it tomorrow and we'll see ☺️